src/System.Linq/src/System/Linq/AnyAll.cs

```diff
@@ -15,6 +16,29 @@ public static bool Any<TSource>(this IEnumerable<TSource> source)
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
             }
 
+            if (source is ICollection<TSource> collectionoft)
+            {
+                return collectionoft.Count != 0;
+            }
+            else if (source is IIListProvider<TSource> listProv)
+            {
+                // Note that this check differs from the corresponding check in
+                // Count (whereas otherwise this method parallels it).  If the count
+                // can't be retrieved cheaply, that likely means we'd need to iterate
+                // through the entire sequence in order to get the count, and in that
+                // case, we'll generally be better off falling through to the logic
+                // below that only enumerates at most a single element.
+                int count = listProv.GetCount(onlyIfCheap: true);
+                if (count >= 0)
+                {
+                    return count != 0;
+                }
+            }
+            else if (source is ICollection collection)
```

Comment:
When will this code path get hit (the non-generic ICollection when the IEnumerable itself is generic)?

Reply:
e.g. `Queue<T>`

Reply:
Interesting. Ty.